File: .leetcode/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start

#include <unordered_map>

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_set<char> occ;
        int l = s.size();
        int rk = -1, ans = 0;
        for (int i = 0; i < l; i++)
        {
            if (i != 0)
                occ.erase(s[i - 1]);
            while (rk + 1 < l && !occ.count(s[rk + 1]))
            {
                occ.insert(s[rk + 1]);
                rk++;
            }
            ans = max(ans, rk - i + 1);
        }
        return ans;
    }
};
// @lc code=end

File: .leetcode/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        n = len(s)
        right, ans = -1, 0
        for i in range(n):
            if i != 0:
                char_set.remove(s[i - 1])
            while right < n - 1 and s[right + 1] not in char_set:
                char_set.add(s[right + 1])
                right += 1
            ans = max(right - i + 1, ans)
        return ans


# @lc code=end
